src/lsm/scan_tree.zig

```diff
@@ -0,0 +1,666 @@
+        table_immutable_cursor: RangeCursor,
+
+        state: union(enum) {
+            idle: void,
+            seeking,
+            fetching: struct {
+                context: Context,
+                callback: Callback,
+                pending_count: u32,
+            },
+        },
+        levels: [constants.lsm_levels]ScanTreeLevel,
+
+        merge_iterator: ?KWayMergeIterator,
+
+        pub fn init(
+            tree: *Tree,
+            buffer: *const ScanBuffer,
+            snapshot: u64,
+            key_min: Key,
+            key_max: Key,
+            direction: Direction,
+        ) ScanTree {
+            assert(key_min <= key_max);
+
+            tree.table_mutable.sort();
+            const table_mutable_cursor = RangeCursor.init(
+                binary_search.binary_search_values_range(
+                    Key,
+                    Value,
+                    key_from_value,
+                    tree.table_mutable.values_used(),
+                    key_min,
+                    key_max,
+                ),
+                direction,
+            );
+
+            const table_immutable_cursor = RangeCursor.init(
+                binary_search.binary_search_values_range(
+                    Key,
+                    Value,
+                    key_from_value,
+                    tree.table_immutable.values_used(),
+                    key_min,
+                    key_max,
+                ),
+                direction,
+            );
+
+            return .{
+                .tree = tree,
+                .buffer = buffer,
+
+                .state = .idle,
+                .snapshot = snapshot,
+                .key_min = key_min,
+                .key_max = key_max,
+                .direction = direction,
+                .table_mutable_cursor = table_mutable_cursor,
+                .table_immutable_cursor = table_immutable_cursor,
+                .levels = undefined,
+                .merge_iterator = null,
+            };
+        }
+
+        pub fn read(scan: *ScanTree, context: Context, callback: Callback) void {
+            switch (scan.state) {
+                .idle => for (&scan.levels, 0..) |*level, i| {
+                    level.* = ScanTreeLevel.init(
+                        scan,
+                        scan.buffer.levels[i],
+                        @as(u8, @intCast(i)),
+                    );
+                    level.next_table();
+                },
+                .seeking => {},
+                .fetching => unreachable,
+            }
+
+            scan.state = .{
+                .fetching = .{
+                    .context = context,
+                    .callback = callback,
+                    .pending_count = 0,
+                },
+            };
+
+            // Track an extra "level" that will finish after the loop.
+            // This allows us to call the callback if there's no more levels to fetch.
+            scan.state.fetching.pending_count += 1;
```

Comment:
Added comments explaining the reason.

It's a workaround for KWayMerging scans, since we don't expose which scan was drained (across all composable scans), we just call `read()` in all of them, so the scan may not have anything to fetch and will finish on `next_tick`.

I want to solve this properly when we implement other kinds of merge (e.g. ZigZagMerge). For now, deferring to `next_tick` doesn't do any harm because some scans will finish asynchronously anyway.